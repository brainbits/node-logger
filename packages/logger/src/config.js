/* eslint-disable import/no-dynamic-require, global-require, no-param-reassign */
import fs from 'fs';
import { join } from 'path';

const defaultConfig = {
    channel: 'unknown',
    levels: [
        'emergency',
        'alert',
        'critical',
        'error',
        'warning',
        'notice',
        'info',
        'debug',
    ],
    maxLevel: 'info',
    outputs: {
        emergency: 'stderr',
        warning: 'stdout',
    },
    timerLevel: 'debug',
    formatter: null,
    plugins: [],
};

function getPackageConfigInPath(path) {
    const packageJsonPath = join(path, '..', 'package.json');

    if (!fs.existsSync(packageJsonPath)) {
        return null;
    }

    const pkg = require(packageJsonPath);

    return { channel: pkg.name || defaultConfig.channel, ...pkg.nodeLogger };
}

function loadConfigFromPackage() {
    const { paths } = process.mainModule;
    const pkgConfig = paths
        .map(path => ({ path, config: getPackageConfigInPath(path) }))
        .find(config => !!config.config);

    if (!pkgConfig) {
        return { path: null, config: {} };
    }

    return pkgConfig;
}

function loadModule(moduleName, path) {
    return require(join(path, moduleName)).default;
}

function loadFormatter(config, path) {
    const { formatter } = config;

    if (typeof formatter === 'function') {
        return formatter;
    }

    if (!path) {
        throw new Error('Formatter must be configured as function unless a package.json is available');
    }

    try {
        return loadModule(formatter, path);
    } catch (error) {
        throw new Error(`Formatter ${formatter} not found in ${path}: ${error.message}`);
    }
}

function loadPlugin(pluginConfig, path) {
    try {
        return loadModule(pluginConfig, path);
    } catch (error) {
        throw new Error(`Plugin ${pluginConfig} not found in ${path}: ${error.message}`);
    }
}

function loadPlugins(config, path) {
    return config
        .plugins
        .map((plugin) => {
            const Plugin = loadPlugin(plugin, path);
            return new Plugin(config);
        });
}

function deepParseEnv(config) {
    const parsedConfig = {};

    if (!config || Array.isArray(config) || typeof config === 'boolean' || typeof config === 'function') {
        return config;
    }

    if (typeof config !== 'object' && typeof config === 'string') {
        const match = config.match(/env\(([^,)]*)(, *(.*))?\)/);

        if (match) {
            if (!process.env[match[1]] && !match[3]) {
                throw new Error(`Env ${match[1]} is not set nor has a fallback!`);
            }

            return process.env[match[1]] || match[3];
        }

        return config;
    }

    Object.entries(config).forEach(([key, value]) => {
        parsedConfig[key] = deepParseEnv(value);
    });

    return parsedConfig;
}

function assertConfig(config) {
    if (!config.formatter) {
        throw new Error('Invalid formatter: No formatter found in configuration');
    }

    if (!Array.isArray(config.levels) || config.levels.length === 0) {
        throw new Error('Invalid levels: No levels were configured');
    }

    if (!config.maxLevel || config.levels.indexOf(config.maxLevel) < 0) {
        throw new Error(`Invalid maxLevel: ${config.maxLevel}`);
    }

    if (!config.timerLevel || config.levels.indexOf(config.timerLevel) < 0) {
        throw new Error(`Invalid timerLevel: ${config.timerLevel}`);
    }

    if (!config.outputs || Object.keys(config.outputs).find(k => !config.levels.includes(k))) {
        throw new Error('Invalid outputs: Outputs can only be configured for existing log levels');
    }

    if (Object.values(config.outputs).find(k => !['stdout', 'stderr'].includes(k))) {
        throw new Error('Invalid outputs: Output must be stdout or stderr');
    }
}

export default function loadConfiguration(config = {}) {
    const packageJsonConfigDefinition = loadConfigFromPackage();
    const merged = deepParseEnv({
        ...defaultConfig,
        ...packageJsonConfigDefinition.config,
        ...config,
    });

    assertConfig(merged);

    return {
        ...merged,
        formatter: loadFormatter(merged, packageJsonConfigDefinition.path),
        plugins: loadPlugins(merged, packageJsonConfigDefinition.path),
    };
}
