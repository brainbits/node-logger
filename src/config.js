/*
 * monoglog: Config that _is_ conform to monolog logging levels.
 */
const { stdout, stderr } = process;


export default {
    levels: {
        emergency: 0,
        alert: 1,
        critical: 2,
        error: 3,
        warning: 4,
        notice: 5,
        info: 6,
        debug: 7,
    },
    outputs: {
        emergency: stderr,
        warning: stdout,
    },
    timerLevel: 'debug',
};
