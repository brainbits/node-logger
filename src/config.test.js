

import config from './config';

describe('Config', () => {
    it('should have the correct levels', () => {
        expect(config.levels).toMatchSnapshot();
    });
    it('should have the correct timer level', () => {
        expect(config.timerLevel).toBe('debug');
    });
    it('should have the correct outputs', () => {
        expect(config.outputs.warning).toEqual(process.stdout);
        expect(config.outputs.emergency).toEqual(process.stderr);
    });
});
