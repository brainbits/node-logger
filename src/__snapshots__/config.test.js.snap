// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Config should have the correct levels 1`] = `
Object {
  "alert": 1,
  "critical": 2,
  "debug": 7,
  "emergency": 0,
  "error": 3,
  "info": 6,
  "notice": 5,
  "warning": 4,
}
`;
